### PR TITLE
PIP-1045-reference-workflow

### PR DESCRIPTION
**Pull request by ottojolanki**

### CODE REVIEW

File: wdl/workflows/references/make_references.wdl [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        BowtieIndex? bowtie_index
+        BwaIndex? bwa_index
+        File? center_sites
+        File? chrom_buckets
+        File? chrom_info
+        File? chrom_sizes
+        File? mappable_regions
+        File reference_fasta
+        Int mappability_kmer_length
+        String machine_size_bwa = "large"
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:25]: Should make ReferenceMachineSizes struct. Can look at structs/dnase for example.

File: wdl/workflows/references/make_references.wdl [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                resources=compute.runtimes[machine_size_center],
+        }
+    }
+
+    File center_sites_output = select_first([center_sites, get_center_sites.center_sites_starch])
+
+    output {
+        BowtieIndex bowtie_index_out = bowtie_index_output
+        BwaIndex bwa_index_out = bwa_index_output
+        File center_sites_out = center_sites_output
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:121]: Should gather these in Hotspot1/2 references. Either https://github.com/ENCODE-DCC/dnase-seq-pipeline/blob/30d8d7b626e5b709931edd651c24bdd8a0ab987c/wdl/structs/dnase.wdl#L22 or its parts: https://github.com/ENCODE-DCC/dnase-seq-pipeline/blob/30d8d7b626e5b709931edd651c24bdd8a0ab987c/wdl/structs/hotspot2.wdl#L4 
And https://github.com/ENCODE-DCC/dnase-seq-pipeline/blob/30d8d7b626e5b709931edd651c24bdd8a0ab987c/wdl/structs/hotspot1.wdl#L4

Comment (keenangraham) [reply wdl/workflows/references/make_references.wdl:121]: We also need something to sit next to dnase.wdl in base directory. Could probably move wdl/workflows/references/make_references.wdl to references.wdl (no make) in base directory.

Comment (ottojolanki) [reply wdl/workflows/references/make_references.wdl]: Opted for HS1, HS2 and IndexedFasta

File: tests/functional/json/test_make_references.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
+    "make_references.mappability_kmer_length": 12,
+    "make_references.reference_fasta": "tests/data/ref/chr22.fa",
+    "make_references.machine_sizes": {
+        "bwa": "small",
+        "fai": "small",
+        "bowtie": "small",
+        "mappable": "small",
+        "chr_buckets": "small",
```
Comment (keenangraham) [inline tests/functional/json/test_make_references.json:9]: Remove chr_buckets from struct.

File: tests/functional/json/test_make_references_bwa_bowtie_defined.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        "sa": "tests/data/ref/chr22.fa.sa"
+    },
+    "make_references.mappability_kmer_length": 12,
+    "make_references.reference_fasta": "tests/data/ref/chr22.fa",
+    "make_references.machine_sizes": {
+        "bwa": "small",
+        "fai": "small",
+        "bowtie": "small",
+        "mappable": "small",
+        "chr_buckets": "small",
```
Comment (keenangraham) [inline tests/functional/json/test_make_references_bwa_bowtie_defined.json:25]: Remove.

File: tests/functional/json/test_make_references_bwa_bowtie_mappable_defined.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    },
+    "make_references.mappability_kmer_length": 12,
+    "make_references.mappable_regions": "tests/data/dnase/references/chr22.K12.mappable_only.bed",
+    "make_references.reference_fasta": "tests/data/ref/chr22.fa",
+    "make_references.machine_sizes": {
+        "bwa": "small",
+        "fai": "small",
+        "bowtie": "small",
+        "mappable": "small",
+        "chr_buckets": "small",
```
Comment (keenangraham) [inline tests/functional/json/test_make_references_bwa_bowtie_mappable_defined.json:26]: Remove.

File: tests/functional/json/test_make_references.json (hunk 3 above)
Comment (keenangraham) [inline tests/functional/json/test_make_references.json:9]: Should change all chr -> chrom to be consistent with e.g. https://github.com/ENCODE-DCC/dnase-seq-pipeline/blob/3fce1ab3fc8fab7abc49e578191bd78ee6540e8e/wdl/structs/hotspot2.wdl#L5

File: wdl/workflows/references/make_references.wdl [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+import "../../subworkflows/build_bowtie_index.wdl" as bowtie
+import "../../subworkflows/build_bwa_index.wdl" as bwa
+import "../../subworkflows/build_fasta_index.wdl" as fai
+import "../../subworkflows/build_mappable_only_bed.wdl" as mappable
+import "../../subworkflows/get_center_sites.wdl" as center
+import "../../subworkflows/get_chrom_info.wdl" as chr_info
+import "../../subworkflows/get_chrom_sizes.wdl" as chr_sizes
+import "../../structs/sizes.wdl"
+
+workflow make_references {
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:13]: make_references -> references

File: wdl/workflows/references/make_references.wdl [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    input {
+        BowtieIndex? bowtie_index
+        BwaIndex? bwa_index
+        File? center_sites
+        File? chrom_info
+        File? chrom_sizes
+        File? mappable_regions
+        File reference_fasta
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:21]: reference_fasta -> fasta?

File: wdl/workflows/references/make_references.wdl [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        Int mappability_kmer_length
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:22]: kmer_length or read_length

File: wdl/workflows/references/make_references.wdl [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        ReferenceMachineSizes machine_sizes = read_json("wdl/default_reference_machine_sizes.json")
+    }
+
+    Machines compute = read_json("wdl/runtimes.json")
+    String genome_name = basename(reference_fasta, ".fa")
+    
+    if (!defined(bwa_index)) {
+        call bwa.build_bwa_index {
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:30]: fasta_for_bwa_index.build_bwa_index

File: wdl/workflows/references/make_references.wdl [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            input:
+                fasta=reference_fasta,
+                resources=compute.runtimes[machine_sizes.bwa],
+        }
+    }
+
+    BwaIndex bwa_index_output = select_first([bwa_index, build_bwa_index.bwa_index])
+
+    call fai.build_fasta_index {
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:39]: fasta_for_fasta_index.build_fasta_index

File: wdl/workflows/references/make_references.wdl [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        input:
+            fasta=reference_fasta,
+            resources=compute.runtimes[machine_sizes.fai],
+    }
+
+    if (!defined(bowtie_index)) {
+        call bowtie.build_bowtie_index {
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:46]: fasta_for_bowtie_index.build_bowtie_index

File: wdl/workflows/references/make_references.wdl [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+            input:
+                fasta=reference_fasta,
+                resources=compute.runtimes[machine_sizes.bowtie]
+        }
+    }
+
+    BowtieIndex bowtie_index_output = select_first([bowtie_index, build_bowtie_index.bowtie_index])
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:53]:
```
select_first([
    bowtie_index,
    build_bowtie_index.bowtie_index
])
```

File: wdl/workflows/references/make_references.wdl [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        call mappable.build_mappable_only_bed {
+            input:
+                bowtie_index=bowtie_index_output,
+                kmer_length=mappability_kmer_length,
+                reference_genome=reference_fasta,
+                resources=compute.runtimes[machine_sizes.mappable],
+        }
+    }
+
+    File mappable_regions_output = select_first([mappable_regions, build_mappable_only_bed.mappable_regions])
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:65]:
```
select_first([
    mappable_regions,
    build_mappable_only_bed.mappable_regions
])
```

File: wdl/workflows/references/make_references.wdl [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+    if (!defined(chrom_sizes)) {
+        call chr_sizes.get_chrom_sizes {
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:68]: fasta_index.get_chrom_sizes

File: wdl/workflows/references/make_references.wdl [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+            input:
+                fai=select_first([build_fasta_index.indexed_fasta.fai]),
+                resources=compute.runtimes[machine_sizes.chr_sizes],
+        }
+    }
+
+    File chrom_sizes_output = select_first([chrom_sizes, get_chrom_sizes.chrom_sizes])
+
+    if (!defined(chrom_info)) {
+        call chr_info.get_chrom_info {
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:78]: chrom_sizes.get_chrom_info

File: wdl/workflows/references/make_references.wdl [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+            input:
+                chrom_sizes=chrom_sizes_output,
+                resources=compute.runtimes[machine_sizes.chr_info],
+        }
+    }
+
+    File chrom_info_output = select_first([chrom_info, get_chrom_info.chrom_info])
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:85]:
```
select_first([
    chrom_info,
    get_chrom_info.chrom_info
])
```

File: wdl/workflows/references/make_references.wdl [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+                resources=compute.runtimes[machine_sizes.mappable],
+        }
+    }
+
+    File mappable_regions_output = select_first([mappable_regions, build_mappable_only_bed.mappable_regions])
+
+    if (!defined(chrom_sizes)) {
+        call chr_sizes.get_chrom_sizes {
+            input:
+                fai=select_first([build_fasta_index.indexed_fasta.fai]),
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:70]:
```
fai=select_first([
    build_fasta_index.indexed_fasta.fai
]),
```

File: wdl/workflows/references/make_references.wdl [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    if (!defined(center_sites)) {
+        call center.get_center_sites {
+            input:
+                chrom_sizes=chrom_sizes_output,
+                mappable_regions=mappable_regions_output,
+                resources=compute.runtimes[machine_sizes.center],
+        }
+    }
+
+    File center_sites_output = select_first([center_sites, get_center_sites.center_sites_starch])
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:96]:
```
select_first([
    center_sites,
    get_center_sites.center_sites_starch
])
```

File: wdl/workflows/references/make_references.wdl [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+
+    output {
+        BwaIndex bwa_index_out = bwa_index_output
+        HotSpot1Reference hotspot1_reference = {
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:100]:
```
object {
    chrom_info: chrom_info_output,
    mappable_regions: mappable_regions_output
}
```

File: wdl/workflows/references/make_references.wdl [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+            "chrom_info": chrom_info_output,
+            "mappable_regions": mappable_regions_output
+        }
+        HotSpot2Reference hotspot2_reference = {
```
Comment (keenangraham) [inline wdl/workflows/references/make_references.wdl:104]: object {}

File: tests/functional/json/test_make_references.json (hunk 3 above)
Comment (keenangraham) [reply tests/functional/json/test_make_references.json:9]: Actually maybe should be named after subworkflows they are calling. `build_fasta_index`, `get_chrom_info`, etc.

Comment (ottojolanki) [reply tests/functional/json/test_make_references.json]: yeah, i like naming according to the subworkflow.

**Pull request merged**